### PR TITLE
EIP-1559 - Elevate gas properties to the Popover, disable submission if errors

### PR DESCRIPTION
**Pull request by darkwing**
Since we need to be able to validate data before submission, we'll elevate the state of gas fro the Display to the Popover.

### CODE REVIEW

File: ui/components/app/edit-gas-display/edit-gas-display.component.js [hunk 1]
```diff
@@ -19,54 +19,49 @@ import AdvancedGasControls from '../advanced-gas-controls/advanced-gas-controls.
...
-    setGasLimit,
-    estimateToUse,
-    setEstimateToUse,
-    estimatedMinimumFiat,
-    estimatedMaximumFiat,
-    isMaxFeeError,
-    isMaxPriorityFeeError,
-    isGasTooLow,
-  } = useGasFeeInputs(defaultEstimateToUse);
+  console.log('defaultEstimateToUse: ', defaultEstimateToUse);
```
Comment (brad-decker) [inline ui/components/app/edit-gas-display/edit-gas-display.component.js:64]: console.log here

**Review by brad-decker — DISMISSED** [review]
forgot a console.log but looks good to me

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 2]
```diff
@@ -55,26 +94,42 @@ export default function EditGasPopover({
     if (!transaction || !mode) {
       closePopover();
     }
+
+    const cancelSpeedUpGas = process.env.SHOW_EIP_1559_UI
+      ? {
+          gasLimit: decimalToHex(gasLimit),
```
Comment (darkwing) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:100]: I need to figure out the root cause of `Cannot convert string to buffer. toBuffer only supports 0x-prefixed hex strings and this string was given: 0x5.100000001` when using this `decimalToHex` function.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 3]
```diff
@@ -55,26 +94,42 @@ export default function EditGasPopover({
     if (!transaction || !mode) {
       closePopover();
     }
+
+    const cancelSpeedUpGas = process.env.SHOW_EIP_1559_UI
```
Comment (brad-decker) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:98]: i'd switch on gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 2 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:100]: Ah, you need to convert these using `decGWEIToHexWEI` from `shared/modules/conversion.util.js` it does two conversions, first from gwei to wei then dec to hex. Not gasLimit though, gasLimit is just dec to hex string.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: shared/constants/gas.js [hunk 4]
```diff
@@ -29,3 +29,12 @@ export const GAS_RECOMMENDATIONS = {
   MEDIUM: 'medium',
   HIGH: 'high',
 };
+
+/**
+ * These represent the different edit modes presented in the UI
+ */
+export const EDIT_GAS_MODES = {
```
Comment (brad-decker) [inline shared/constants/gas.js:36]: [REQ] These are UI specific, they don't need to be in the shared folder. I'd recommend leaving these declared in the EditPopover as they primarily modify the behavior of this component and should likely only be imported when also importing EditPopover

File: ui/components/app/edit-gas-display/edit-gas-display.component.js [hunk 5]
```diff
@@ -19,55 +22,49 @@ import AdvancedGasControls from '../advanced-gas-controls/advanced-gas-controls.
...
 
 import { I18nContext } from '../../../contexts/i18n';
-import { useGasFeeInputs } from '../../../hooks/useGasFeeInputs';
 
 export default function EditGasDisplay({
-  alwaysShowForm,
-  type,
-  showEducationButton,
+  alwaysShowForm = false,
+  type = 'customize-gas',
```
Comment (brad-decker) [inline ui/components/app/edit-gas-display/edit-gas-display.component.js:28]: I think the 'mode' prop removes our need to have the 'type' prop?

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 6]
```diff
@@ -55,26 +93,43 @@ export default function EditGasPopover({
     if (!transaction || !mode) {
       closePopover();
     }
+
+    const cancelSpeedUpGas =
+      gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET
+        ? {
+            gasLimit: decGWEIToHexWEI(gasLimit),
```
Comment (brad-decker) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:100]: gasLimit is not in gwei, it needs to be just a straight conversion to hex. decimalToHex.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 7]
```diff
@@ -55,26 +93,43 @@ export default function EditGasPopover({
     if (!transaction || !mode) {
       closePopover();
     }
+
+    const cancelSpeedUpGas =
```
Comment (brad-decker) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:97]: rename to newGasSettings and use it for all three 'modes' below

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 8]
```diff
@@ -55,26 +93,43 @@ export default function EditGasPopover({
...
         break;
-      case EDIT_GAS_MODE.MODIFY_IN_PLACE:
+      case EDIT_GAS_MODES.MODIFY_IN_PLACE:
         dispatch(
           updateTransaction({
             ...transaction,
-            txParams: { ...transaction.txParams /** ...newGasSettings */ },
+            txParams: {
+              ...transaction.txParams,
+              ...(process.env.SHOW_EIP_1559_UI
```
Comment (brad-decker) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:122]: this is essentially the same as the cancelSpeedUpGas, so you can use it here too, just add the gas key manually. it won't hurt anything to add both gasLimit and gas (only one will be used)

**Review by brad-decker — CHANGES_REQUESTED** [review]
Looks good just a few little things.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 8 above)
Comment (darkwing) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js]: Good to know!  Updated!

File: shared/constants/gas.js (hunk 4 above)
Comment (darkwing) [reply shared/constants/gas.js:36]: Pushing back on this slightly; `GAS_RECOMMENDATIONS` is also a shared constant which is only used in the UI.  I think having it here will be really helpful here.

Comment (brad-decker) [reply shared/constants/gas.js:36]: I didn't comment on the addition of `GAS_RECOMMENDATIONS` because we'll need that in the background now, too. We'll be using it when we set default gas estimations on incoming transactions from dapps. 

I understand the desire to place this here, but it is important that we keep the shared folder strict to its purpose of storing code shared by both `app/` and `ui/` folders to keep the module boundary line clearly defined.

**Review by brad-decker — DISMISSED** [review]
LGTM with the exception of the location of the constant -- but that can either be moved here or we can wait until we are sure its not used in the background before we relocate it at some point in the future.

**Review by brad-decker — DISMISSED** [review]

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 9]
```diff
@@ -55,26 +96,36 @@ export default function EditGasPopover({
     if (!transaction || !mode) {
       closePopover();
     }
+
+    const newGasSettings =
+      gasEstimateType === GAS_ESTIMATE_TYPES.FEE_MARKET
+        ? {
+            gas: decimalToHex(gasLimit),
+            gasLimit: decimalToHex(gasLimit),
```
Comment (Gudahtt) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: This seems like a mistake :thinking: 
```suggestion
```

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 10]
```diff
@@ -16,25 +28,54 @@ import {
...
+    gasLimit,
+    setGasLimit,
+    estimateToUse,
+    setEstimateToUse,
+    estimatedMinimumFiat,
+    estimatedMaximumFiat,
+    isMaxFeeError,
+    isMaxPriorityFeeError,
+    isGasTooLow,
+  } = useGasFeeInputs(defaultEstimateToUse);
```
Comment (Gudahtt) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:77]: I don't have the proper context here, but passing this many props around seems like a red flag to me. This is quite difficult to understand. It was overly complex before too when it was isolated to the "edit-gas-display" component, and now it's worse because the complexity has found its way into a component API now as well. I don't see how we could make a component with over a dozen props easily understood or used.

A better way to handle UI complexity would be to track UI state in a Redux store, and grab it directly from Redux wherever a value is needed. Then derived state can be calculated using selectors. Then hopefully the UI logic would be just that - UI logic, not cluttered with the rules of how editing gas works.

Anyway - that's not something that was introduced here, nor can it be fixed here in this PR. Just commenting on it now that I've noticed it.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (darkwing) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js]: Good shout, removed!

File: ui/components/app/edit-gas-display/edit-gas-display.component.js [hunk 11]
```diff
@@ -141,31 +137,33 @@ export default function EditGasDisplay({
...
-                label: t('editGasHigh'),
-                recommended: defaultEstimateToUse === GAS_RECOMMENDATIONS.HIGH,
-              },
-            ]}
-            selectedValue={estimateToUse}
-            onChange={setEstimateToUse}
-          />
-        )}
+        {!requireDappAcknowledgement &&
+          ![EDIT_GAS_MODES.SPEED_UP, EDIT_GAS_MODES.CANCEL].includes(mode) && (
```
Comment (Gudahtt) [inline ui/components/app/edit-gas-display/edit-gas-display.component.js:141]: No recommendations for speed up or cancellation?

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js [hunk 12]
```diff
@@ -83,11 +133,29 @@ export default function EditGasPopover({
...
+    gasLimit,
+    gasPrice,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    gasEstimateType,
+  ]);
+
+  let title = t('editGasTitle');
+  if (popoverTitle) {
+    title = popoverTitle;
```
Comment (Gudahtt) [inline ui/components/app/edit-gas-popover/edit-gas-popover.component.js:150]: Just noting that the title priorities have changed here as well. Previously the education title would be shown over a custom title, but no longer.

This seems more appropriate though I think :thinking:

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 10 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:77]: Our goal is to get to a working solution in as iterative a process as possible. We've discussed moving much of this logic into redux but that will happen at a later point when we have a clearer picture of where we are heading.

Most values come directly from the background state, and are computed using conversion methods and other hooks (useCurrencyDisplay)

Other values come from the transaction being modified which we used to set in state and access in redux, or passed in through modal or popover.

So I'd like to take more time to think about how we integrate this into redux

**Review by Gudahtt — DISMISSED** [review]
LGTM! Seems equivalent except up one layer, except that now the new gas settings are included with the speedup/cancellation call, and the form is disabled when there is an error. And there were a few more functional changes that seemed probably intentional, but I left comments so that someone can double-check.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js]: Need this, its part of the new options bag for speed up and cancel methods, and transactionFactory from ethereumjs/tx looks for gasLimit too. We need this line for now we can figure out if its redundant later.

Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js]: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L557-L625

File: ui/components/app/edit-gas-display/edit-gas-display.component.js (hunk 11 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-display/edit-gas-display.component.js:141]: yeah i believe we should be showing recommendations for these. Previously we would just default the fields to 110% of the previous value but there was no warning for going under that. Now we're going to use present estimations but warn if the value is less than 110% of previous value. That work doesn't have to be done here but I'm also curious about this logic change.

**Comment by darkwing** [comment]
Restored `gasLimit`

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 10 above)
Comment (Gudahtt) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:77]: It's far easier to iterate with smaller and simpler APIs than it is with enormously complex ones. I don't think an iterative approach is in conflict with good API design.

When I suggested Redux I hadn't looked closely enough to be honest. I see now that some of this is input field state, and that definitely makes sense to be at the component level. So maybe some of this is best here, and maybe some is best as utility functions or selectors or as hooks broken out into more logical chunks. Or maybe some should be in Redux, I don't know. But I'm sure we can come up with a more readable abstraction than this enormous list of props.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (Gudahtt) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: That's disappointing. The type definition even states that there is no gas limit: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L56

File: ui/components/app/edit-gas-display/edit-gas-display.component.js (hunk 11 above)
Comment (darkwing) [reply ui/components/app/edit-gas-display/edit-gas-display.component.js:141]: I can restore it, but they aren't present in the Figma file:  https://www.figma.com/file/NXvUKcZ0Ocu1GPcT7VxRMx/1559---Gas-Controls?node-id=643%3A5

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (Gudahtt) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: What is the intended difference between `gas` and `gasLimit`? I'm finding this very hard to follow. `generateNewGasParams` for example uses the truthiness of `gasLimit` to decide whether to ensure `gas` was set, but it doesn't use `gasLimit` directly. It doesn't seem to be explained in any comments either :disappointed:

Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: A big reason for the introduction of `gasLimit` is because `@ethereumjs/tx` has no mention of a `gas` field in its [`txData` type](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/types.ts#L112-L163), but does specify `gasLimit`.

`generateNewGasParams` has some flaws in its types defined and I did not do a good enough job writing down my internal thoughts/strategy. I had assumed that the usage of `gasLimit` by `@ethereumjs/tx` was a strong signal that we should also be using it instead of `gas` but in the midst of that discovered that the JSON RPC spec does not use `gasLimit`. It doesn't matter because we only use that package to generate a serialized tx that abstracts away field names. 

My thought was to consolidate to using one field name, personally gasLimit makes the most sense to me and is more descriptive than gas, but I wanted to take the time to understand the impact that would have. You'll see in the signTransaction method I added gasLimit to the txParams before signing to match the type. 

All of this to say, I made some choices during individual units of work that led us here. I'll make it a priority to update the documentation of the new method and the create speed up/cancel methods. I don't think this should be addressed here. Happy to have further conversation about this.

File: ui/components/app/edit-gas-display/edit-gas-display.component.js (hunk 11 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-display/edit-gas-display.component.js:141]: talked on the sync up this morning and they do not get recommendations right now we may restore at a later time.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (Gudahtt) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: Hmm. So just to be clear - both properties have the same intended purpose? They're just duplicated for the aforementioned reasons.

I'm asking because your earlier comment made me think that both were needed for speed up and cancellation specifically, but not for other transactions.

**Review by brad-decker — APPROVED** [review]

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 9 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:104]: yes, they are redundant fields, they ultimately set the `gas` value on txParams.

File: ui/components/app/edit-gas-popover/edit-gas-popover.component.js (hunk 10 above)
Comment (brad-decker) [reply ui/components/app/edit-gas-popover/edit-gas-popover.component.js:77]: We know we can come up with something, as well. I don't disagree with your sentiment but we're doing the best we can with our current knowledge of the problem, in the time constraints we are under without breaking anything that is currently in place for transactions today (before the fork). The consolidation of logic into a hook was a purposeful step meant to get us into a place where abstracting that into a redux store (or wherever) would be easier in the future.

**Review by danjm — APPROVED** [review]
